### PR TITLE
Generate odra::Module implementation

### LINKED ISSUE #219: Add `odra-macros` crate
Migrate `odra-macros` from the playground repo and provide a macro `odra::module2`. It should produce code that works with `mock-vm` at first.

### PR DESCRIPTION
**Pull request by kpob**
Macro implementation
module macro is applicable to `struct` and `impl` block
Examples update - remove manually written code

### CODE REVIEW

**Review by kubaplas — APPROVED** [review]

File: examples2/src/counter_pack.rs [hunk 1]
```diff
@@ -5,6 +5,7 @@ use odra::Mapping;
 use odra::Module;
 use odra::ModuleWrapper;
 
+#[odra_macros::module]
```
Comment (zie1ony) [inline examples2/src/counter_pack.rs:8]: Why not `odra::module`?

File: odra-macros/src/ast/host_ref_item.rs [hunk 2]
```diff
@@ -198,7 +198,7 @@ mod ref_item_tests {
                     self.env.last_call().contract_last_call(self.address)
                 }
 
-                pub fn try_total_supply(&self) -> Result<U256, OdraError> {
+                pub fn try_total_supply(&self) -> Result<U256, odra::OdraError> {
```
Comment (zie1ony) [inline odra-macros/src/ast/host_ref_item.rs:201]: I just noticed this. Do we need `try_` for `&self` methods?

File: odra-macros/src/ir/mod.rs [hunk 3]
```diff
@@ -1,26 +1,109 @@
...
+    pub fn field_names(&self) -> Result<Vec<syn::Ident>, syn::Error> {
+        utils::syn::struct_fields_ident(&self.code)
+    }
+
+    pub fn module_ident(&self) -> syn::Ident {
+        utils::syn::ident_from_struct(&self.code)
+    }
+
+    pub fn module_mod_ident(&self) -> syn::Ident {
+        format_ident!(
```
Comment (zie1ony) [inline odra-macros/src/ir/mod.rs:42]: I'd move it to `utils`.

**Review by zie1ony — APPROVED** [review]

File: examples2/src/counter_pack.rs (hunk 1 above)
Comment (kpob) [reply examples2/src/counter_pack.rs:8]: Will be, must be reexported in `odra` first.

File: odra-macros/src/ast/host_ref_item.rs (hunk 2 above)
Comment (kpob) [reply odra-macros/src/ast/host_ref_item.rs:201]: Yes. First, for consistency. Second, a `&self` function may have guards and fail for some reason. But testing your code you can use a regular function if you are sure it does not fail.

File: examples2/src/counter_pack.rs (hunk 1 above)
Comment (zie1ony) [reply examples2/src/counter_pack.rs:8]: You should do it in this PR.

Comment (kpob) [reply examples2/src/counter_pack.rs:8]: I could but it does not matter, the feature still is not completed, and there will be a few more prs related to macros, and the release branch is under development.

Comment (zie1ony) [reply examples2/src/counter_pack.rs:8]: up to you

**Pull request merged**